Add Windows PV tools installer tests

dinhngtu opened the pull request:
This PR adds tests for the new XCP-ng 9.x Windows PV Tools installer.
It tests various actions of the installer (install, uninstall, upgrade) under various conditions (with XCP-ng v8 tools, with Citrix tools, with Windows Update drivers, etc.)
It adds a dependency on 2 new ISO images, the structures of which are documented below:

```
guest-tools-win.iso
├───package
│   └───XenClean
│       └───x64
│           └───bin
└───testsign
```

```
other-guest-tools-win.iso
├───citrix-9.4.0
│   ├───XenBus
│   │   ├───x64
│   │   └───x86
│   ├───XenIface
│   │   ├───x64
│   │   └───x86
│   ├───XenNet
│   │   ├───x64
│   │   └───x86
│   ├───XenVbd
│   │   ├───x64
│   │   └───x86
│   └───XenVif
│       ├───x64
│       └───x86
├───xcp-ng-8.2.2.200
│   ├───xenbus
│   │   ├───x64
│   │   └───x86
│   ├───xeniface
│   │   ├───x64
│   │   └───x86
│   ├───xennet
│   │   ├───x64
│   │   └───x86
│   ├───xenvbd
│   │   ├───x64
│   │   └───x86
│   └───xenvif
│       ├───x64
│       └───x86
└───xcp-ng-9.0.9000
    ├───package
    │   └───XenClean
    │       └───x64
    │           └───bin
    └───testsign
```

I have run these tests on Windows Server 2016, 2019, 2022 and 2025. I used customized images with testsigning enabled.

stormi commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+
+@pytest.fixture
+def vm_instance(vm_prepared: tuple[VM, Snapshot]):
+    (vm, snapshot) = vm_prepared
```
AFAIK the usual convention is to omit the optional parentheses of the tuple when unpacking to variables (several instances of this in the file).

```suggestion
    vm, snapshot = vm_prepared
```

stormi commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+
+
+@pytest.fixture
+def vm_instance(vm_prepared: tuple[VM, Snapshot]):
```
```suggestion
def running_unsealed_windows_vm(vm_prepared: tuple[VM, Snapshot]):
```

stormi commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+@pytest.mark.usefixtures("windows_vm")
+class TestGuestToolsWindows:
+    @pytest.fixture(scope="class")
+    def vm_install(self, vm_prepared):
```
I had to read the fixture definition to understand what `install` means in this context, so here also I'd go for a longer, more explicit name.

For example `vm_with_our_tools`?

stormi commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+
+    def test_drivers_detected(self, vm_install: VM):
+        vm = vm_install
+        assert vm.param_get("PV-drivers-detected")
```
We don't test the management agent. I suppose we'll expand the test later when we provide it with the drivers?

stormi commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+from lib.vm import VM
+from . import *
+
+
```
We describe the requirements on top of test files : pools, hosts, VMs, additional resources, specific configurations...

stormi commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+    @pytest.fixture
+    def vm_install_wu_drivers(self, vm_prepared: VM):
+        (vm, snapshot) = vm_prepared
+        vm.param_set("has-vendor-device", True)
```
This makes me think you should probably verify that it's false, in the initial fixture. We attempt to detect failure causes as early as possible, as this is usually easier to understand when a test fails. Here, if someone attemps to run the test on a badly made XVA, or on an existing VM without realizing that the setting had to be false for the tests where we install our tools, the resulting error might not be obvious.

stormi commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+        vm.param_set("has-vendor-device", True)
+        vm.start()
+        wait_for_guest_start(vm)
+        install_other_drivers(
+            vm,
+            "citrix-9.4.0",
+            is_msi=False,
+        )
```
Does this install the drivers via Windows Update?

dinhngtu replied:
It installs Citrix drivers with the vendor device enabled to mimic situations where PV drivers are installed via Windows Update. I'll rename the fixture.

stormi commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+        return vm
+
+    @pytest.fixture
+    def vm_install_wu_drivers(self, vm_prepared: VM):
```
It's not obvious that "wu" means "Windows Update". Expand the acronym or add a comment?

stormi commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+    )
+    wait_for(vm.is_halted, "Shutdown VM")
+    snapshot = vm.snapshot()
+    yield (vm, snapshot)
```
We've done such fixtures in the past, and this works well. However, restoring a snapshot is a rather costly operation, and it's possible that we would get better performance by using cloned VMs instead (usually very fast thanks to copy-on-write). And this simplifies the `yield` part as you just need to yield the cloned VM.

dinhngtu replied:
I've tried that, but came back to the snapshot approach since there's no need to rediscover VM IP after every run. I'm also not sure if deleting a cloned VM is faster than restoring from a snapshot since both are COW-based anyway.

stormi commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+        )
+        assert exitcode == ERROR_INSTALL_FAILURE
+
+    def test_install_with_xcpng_v8_tools(self, vm_install_xcpng_v8_tools: VM):
+        exitcode = install_cert_and_tools(
+            vm_install_xcpng_v8_tools, PowerAction.Nothing, check=False
+        )
+        assert exitcode == ERROR_INSTALL_FAILURE
```
Do you also envision tests where we start with either of the old v8 tools, xcpng_v9_tools, old citrix tools, recent citrix tools, remove them and replace them with the latest we built?

dinhngtu replied:
Some tools (e.g. xcp-ng v8) don't uninstall cleanly and crash the VM. I'll try to add scenarios where some xenbus/etc. drivers are already installed (without the management packages).

stormi replied:
Could we use the clean-up tool? This would allow to test it in the process, too.

stormi reviewed (commented):
This will be a nice addition to the tests. Here are my review comments.

dinhngtu commented on tests/guest-tools/win/__init__.py:
```diff
@@ -0,0 +1,123 @@
+def try_get_and_store_ip_serial(vm: VM, timeout: int):
+    domid = vm.param_get("dom-id")
+    logging.debug(f"Domain ID {domid}")
+    command = [
+        "xl",
+        "console",
+        "-t",
+        "serial",
+        domid,
+        "|",
+        "grep",
+        r"~xcp-ng-tests~.*~end~",
+        "|",
+        "head",
+        "-n",
+        "1",
+    ]
```
`commands.py _ssh()` does a `command = " ".join(cmd)` unconditionally, which breaks if `cmd` is a string. Should I add a check of `if cmd is not str` there?

stormi replied:
Yes, I suppose we could check the type.

dinhngtu commented on tests/guest-tools/win/__init__.py:
```diff
@@ -0,0 +1,123 @@
+        raise Exception(f"Unknown problem status {output}")
+
+
+def run_command_powershell(vm: VM, program: str, args: List[str]):
```
It uses `Start-Process` to run a single executable, and cannot run a script unlike `powershell.exe -encodedcommand`. I'll add a separate `vm.execute_powershell_script` to this effect.

dinhngtu commented on tests/guest-tools/win/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    )
+
+
+def start_background_powershell(vm: VM, cmd: str):
```
The main motivation was to retrieve exit codes higher than 255 (which ssh didn't like) and to avoid Git Bash's path escaping behavior, which undesirably transforms arguments (e.g. `/x` -> `x:/`)

dinhngtu commented on tests/guest-tools/win/test_guest_tools_win.py:
```diff
@@ -0,0 +1,346 @@
+
+
+def install_cert_and_tools(vm: VM, action: PowerAction, check: bool = True):
+    vm.insert_cd("guest-tools-win.iso")
```
Would the Windows tools be tested as part of the XCP-ng package or separately from it? If it's the latter then IMO it's better to download the guest ISO like the other-tools one.

benjamreis reviewed (approved):
LGTM! :+1:

dinhngtu commented:
I'm doing some refactoring to make testing multiple guest tools/other tools versions a bit easier. Currently running final tests.

dinhngtu commented:
I've refactored the tests as following:

- The ISO files of our test tools and other tools could be configured in `data.py` (see `WIN_GUEST_TOOLS_ISOS` and `OTHER_GUEST_TOOLS_ISO`).
- Each ISO configuration now has a `download` parameter that lets the test use an included ISO file or download it from a HTTP location into a local ISO SR.
- Each configuration in `WIN_GUEST_TOOLS_ISOS` specifies a ISO location, whether it requires testsigning, MSI file name, etc.
- We can configure the contents of our other tools ISO file via `OTHER_GUEST_TOOLS` using a similar method.
- All guest and other tools versions now share the same parameterized tests.
- I've split guest-tools/win into several separate module files.
- XenClean tests have been added.

stormi reviewed (commented):
Looks good. Just a few minor comments.

I will delay my approval until we have setup the tests in our CI so that they run once merged on master.

gduperrey commented on jobs.py:
```diff
@@ -350,6 +350,19 @@
+    "tools-windows": {
+        "description": "tests our windows guest tools on a variety of VMs",
+        "requirements": [
+            "A pool >= 8.3. One host is enough.",
```
Does it work in 8.2?

dinhngtu replied:
Yes, it should work.

dinhngtu commented:
I've added some tools and scripts useful for reducing Windows image sizes.

dinhngtu commented:
I've switched the ISO download code to use the existing NFS ISO SR config.

dinhngtu commented:
Changes since last time:
- Reworked `OTHER_GUEST_TOOLS is_msi` property into a nullable enum with values `msi/inf/None`. Separate the "vendor/Windows Update" case and the "Citrix inf" case since these are actually 2 different cases.
- Added several hacks to work around edge cases that caused test failures (e.g. CD insertion hang, CD drive not detected).
- Added the commit "lib/host: Allow specifying default SR UUID" to ease running tests on specific hardware (e.g. particular storage types).
- Added `scripts/guests/windows/install-drivers.ps1`.
- Refactor Windows shutdown command into a constant.

Upgrades are marked as xfail since they require multiple reboots, which I haven't yet taken into account.

gduperrey commented:
CI:
* 8.2: :white_check_mark:
* 8.3: :white_check_mark:

stormi commented:
Thanks for a good PR!
